calico_node/tests/st/config/test_felix_config.py

```diff
@@ -0,0 +1,463 @@
+                    self.open_port(self.host1, port, k)
+
+        for k in failsafe:
+            for port in failsafe[k]:
+                self.check_access_retry(expect_access, self.host2, self.host1, port, k)
+
+    @staticmethod
+    def open_port(host, port, protocol="tcp"):
+        protocol_opt = "-u" if protocol == "udp" else ""
+        cmd = "nc -l %s -p %s -e /bin/sh" % (protocol_opt, port)
+        host.execute(cmd, raise_exception_on_failure=True, daemon_mode=True)
+
+    @staticmethod
+    def close_port(host, port, protocol="tcp"):
+        protocol_opt = "-u" if protocol == "udp" else ""
+        cmd = "nc -l %s -p %s -e /bin/sh" % (protocol_opt, port)
+        host.execute(cmd, raise_exception_on_failure=True, daemon_mode=True)
+
+    def check_access_retry(self, expect_access, host_src, host_target, port, protocol="tcp"):
+        msg = (" from %s to %s:%s %s, expect access %s" %
+               (host_src.ip, host_target.ip, port, protocol, expect_access))
+        for retry in range(3):
+            passed = self.check_access(expect_access, host_src, host_target, port, protocol)
+            if passed:
+                _log.info("check access success!%s", msg)
+                return
+            else:
+                if not expect_access:
+                    # If expect is False but result is True. The port has been
+                    # closed. We need open it again for next round testing.
+                    # If expect is True but result is False, the port is still open.
+                    _log.info("Reopen port for target host.")
+                    self.open_port(host_target, port, protocol)
+                _log.exception("retry [%s] check access failed!%s",
+                               retry, msg)
+                time.sleep(1)
+        self.fail(("check access failed!%s", msg))
+
+    @staticmethod
+    def check_access(expect_access, host_src, host_target, port, protocol="tcp"):
+        protocol_opt = "-u" if protocol == "udp" else ""
+        cmd = "echo \"hostname && exit\" | timeout -t 3 nc %s -w 1 %s %s" % \
+              (protocol_opt, host_target.ip, port)
+        remote_hostname = host_src.execute(cmd, raise_exception_on_failure=False)
+        hostname = host_target.execute("hostname")
+        _log.info("check access from %s to %s:%s %s, result is |%s|, "
+                  "target hostname |%s|, expect access %s",
+                  host_src.ip, host_target.ip, port, protocol,
+                  remote_hostname, hostname, expect_access)
+
+        # If port is 179, we are talking to bird.
+        # If bird returns with a valid string, this means connection is successful.
+        if port == 179 and remote_hostname is not None and len(remote_hostname) > 0:
+            remote_hostname = hostname
+
+        result = (hostname == remote_hostname)
+        return result == expect_access
+
+    def remove_host_endpoint(self):
+        self.host1.delete_all_resource("hostEndpoint")
+
```

Comment:
Spurious whitespace.  I think there should only be one empty line at the bottom of the file.